apm-agent-plugins/apm-apache-httpclient-plugin/src/test/java/co/elastic/apm/agent/httpclient/ApacheHttpAsyncClientInstrumentationTest.java

```diff
@@ -0,0 +1,78 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.httpclient;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.concurrent.FutureCallback;
+import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
+import org.apache.http.impl.nio.client.HttpAsyncClients;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
+public class ApacheHttpAsyncClientInstrumentationTest extends AbstractHttpClientInstrumentationTest {
+
+    private static CloseableHttpAsyncClient client;
+
+    @BeforeClass
+    public static void setUp() {
+        client = HttpAsyncClients.createDefault();
+        client.start();
+    }
+
+    @AfterClass
+    public static void tearDown() throws IOException {
+        client.close();
+    }
+
+    @Override
+    protected void performGet(String path) throws Exception {
+        final CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();
+
+        RequestConfig requestConfig = RequestConfig.custom()
+            .setCircularRedirectsAllowed(true)
+            .build();
+        HttpClientContext httpClientContext = HttpClientContext.create();
+        httpClientContext.setRequestConfig(requestConfig);
+        client.execute(new HttpGet(path), httpClientContext, new FutureCallback<>() {
+            @Override
+            public void completed(HttpResponse result) {
+                responseFuture.complete(result);
+            }
+
+            @Override
+            public void failed(Exception ex) {
+                responseFuture.completeExceptionally(ex);
+            }
+
+            @Override
+            public void cancelled() {
+                responseFuture.cancel(true);
+            }
+        });
+
+        responseFuture.get().getStatusLine().getStatusCode();
```

Comment:
Is it required to get the status line in order to wait that the request is actually executed? Or is that a leftover from debugging?

Reply:
No, only the `get` is important here... I can remove